### PR TITLE
semv and rancher_release label-issues

### PR DESCRIPTION
**Pull request by brooksn**
- Adds a package for parsing semantic versions, and for testing a semantic version range.
- Adds a command semv
- Adds the label-issues subcommand to rancher_release. 

### semv

Usage:
```sh
semv parse v1.2.3-rc1 -o go-template="{{.Major}}.{{.Minor}}.{{.Patch}}"
# 1.2.3-rc1
semv parse v1.2.3-rc1 -o go-template="{{.Prerelease}}"
# -rc1
semv test v1.x v1.2.3
# exit 0
semv test v1.1 v1.2.3
# exit 1
```

### rancher_release label-issues

When a rancher release candidate is ready to test, this command will update all open issues in an affected milestone with the tag `zube: Waiting for RC` to instead have the tag `zube: To Test`, and will add a comment stating which RC with which the issue should be validated.

Usage:
```sh
rancher_release label-issues -t v2.8.8-rc1 --dry-run

Updating 3 issues
#8 Issue A (v2.8.x)
  [zube: Waiting for RC] → [zube: To Test] 
#7 Issue B (v2.8.x)
  [zube: Waiting for RC] → [zube: To Test] 
#6 Issue C (v2.8.x)
  [zube: Waiting for RC] → [zube: To Test] 
```

### CODE REVIEW

File: cmd/rancher_release/label_issues.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:     "tag",
+				Usage:    "release tag to validate images",
+				Required: true,
+			},
+
+			cli.BoolFlag{
+				Name:     "dry-run",
+				Usage:    "the newly created branch won't be pushed to remote and the PR won't be created",
```
Comment (briandowns) [inline cmd/rancher_release/label_issues.go:29]: Is this accurate?

File: cmd/rancher_release/label_issues.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+// labelIssuesWaitingForRC updates issues with the "Waiting for RC" label
+// to "To Test" if the issue's milestone matches the tag, and creates a comment
+func labelIssuesWaitingForRC(c *cli.Context) error {
+	ctx := context.Background()
+	tag := c.String("tag")
+	dryRun := c.BoolT("dry-run")
+	githubToken := os.Getenv("GITHUB_TOKEN")
+	repo := "rancher"
```
Comment (briandowns) [inline cmd/rancher_release/label_issues.go:44]: Can these 4 lines be const?

File: cmd/rancher_release/label_issues.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/google/go-github/v39/github"
+	"github.com/rancher/ecm-distro-tools/release/semver"
+
```
Comment (briandowns) [inline cmd/rancher_release/label_issues.go:11]: Remove blank line and  reformat imports.

File: release/semver/semver.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
+package semver
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+
+	"github.com/rancher/ecm-distro-tools/types"
+
```
Comment (briandowns) [inline release/semver/semver.go:9]: Remove blank line and reformat.

File: cmd/semv/main.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	"fmt"
+	"html/template"
+	"os"
+	"strings"
+
+	"github.com/rancher/ecm-distro-tools/release/semver"
+	"gopkg.in/yaml.v3"
+)
+
+func Format(v *semver.Version, format string) (string, error) {
```
Comment (briandowns) [inline cmd/semv/main.go:17]: Can you move this right above main or right below it so the boilerplate is at the top?

```suggestion
func format(v *semver.Version, format string) (string, error) {
```

File: cmd/semv/main.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	version string
+	gitSHA  string
+)
+
+const usage = `version: %s
+Usage: %[2]s [-test] [-parse]
+Options:
+    -h            help
+    -v            show version and exit
+    -test         test a complete version against a semantic version pattern
```
Comment (briandowns) [inline cmd/semv/main.go:68]: Can we have these flags as `-t` and below, `-p`?

File: release/semver/semver.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	Prerelease string
+	Build      string
+}
+
+type Version struct {
+	Version    string
+	Major      int
+	Minor      int
+	Patch      int
+	Prerelease string
```
Comment (briandowns) [inline release/semver/semver.go:27]: Can we string the hyphen from this field?

**Comment by briandowns** [comment]
Can we rebase this before review again?

File: cmd/rancher_release/label_issues.go (hunk 2 above)
Comment (brooksn) [reply cmd/rancher_release/label_issues.go:50]: Made const

File: cmd/rancher_release/label_issues.go (hunk 3 above)
Comment (brooksn) [reply cmd/rancher_release/label_issues.go]: Formatted

File: release/semver/semver.go (hunk 4 above)
Comment (brooksn) [reply release/semver/semver.go]: Formatted

File: cmd/semv/main.go (hunk 5 above)
Comment (brooksn) [reply cmd/semv/main.go]: Moved this lower and unexported

File: cmd/semv/main.go (hunk 6 above)
Comment (brooksn) [reply cmd/semv/main.go]: Refactored to use urfave with subcommands parse and test

File: release/semver/semver.go (hunk 7 above)
Comment (brooksn) [reply release/semver/semver.go]: Hyphens no longer displayed

**Review by briandowns — APPROVED** [review]

**Review by Tashima42 — APPROVED** [review]

File: cmd/semv/README.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# semv
+
+Parse, test and compare semantic versions
+
+### Examples
+
+```sh
+semv -test v1.1.x v1.1.1-rc1
```
Comment (Tashima42) [inline cmd/semv/README.md:8]:
```suggestion
semv test v1.1.x v1.1.1-rc1
```

File: cmd/semv/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+```
+
+```sh
+semv -is-prerelease v1.1.1-rc1
+```
```
Comment (Tashima42) [inline cmd/semv/README.md:13]:
```suggestion
```

File: cmd/semv/README.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+```sh
+semv -parse v1.1.1-rc1 -o
```
Comment (Tashima42) [inline cmd/semv/README.md:16]:
```suggestion
semv parse v1.1.1-rc1 -o 
```

**Review by Tashima42 — CHANGES_REQUESTED** [review]

**Review by Tashima42 — APPROVED** [review]

**Pull request merged**